dpar/src/lookup.rs

```diff
@@ -0,0 +1,380 @@
+/// let frozen: FreezableLookup<String> = ::serde_yaml::from_str(&serialized)
+///   .expect("Deserialization of transition lookup failed");
+/// assert!(frozen.is_frozen());
+/// ```
+#[derive(Debug, Deserialize, Eq, PartialEq)]
+pub struct FreezableLookup<T>(FreezableLookupInner<T>)
+where
+    T: Eq + Hash;
+
+impl<T> FreezableLookup<T>
+where
+    T: Eq + Hash,
+{
+    /// Check whether the table is frozen.
+    pub fn is_frozen(&self) -> bool {
+        use self::FreezableLookupInner::*;
+
+        match &self.0 {
+            Fresh(_) => false,
+            Frozen(_) => true,
+        }
+    }
+}
+
+impl<T> FreezableLookup<T>
+where
+    T: Eq + Hash,
+{
+    /// Freeze a transition table.
+    ///
+    /// This results in a table that returns the identifier `0` for
+    /// unknown values.
+    ///
+    /// ```
+    /// use dpar::lookup::{FreezableLookup, LookupValue};
+    ///
+    /// let lookup = FreezableLookup::default();
+    /// assert_eq!(lookup.lookup("hello"), 1);
+    /// let frozen = lookup.freeze();
+    /// assert_eq!(frozen.lookup("hello"), 1);
+    /// assert_eq!(frozen.lookup("world"), 0);
+    /// ```
+    pub fn freeze(self) -> Self {
+        use self::FreezableLookupInner::*;
+
+        let inner = match self.0 {
+            Fresh(cell) => Frozen(cell.into_inner()),
+            frozen => frozen,
+        };
+
+        FreezableLookup(inner)
+    }
+
+    /// Thaw a transition table.
+    ///
+    /// This results in a table that adds unknown values when they
+    /// are looked up.
+    ///
+    /// ```
+    /// use dpar::lookup::{FreezableLookup, LookupValue};
+    ///
+    /// let lookup = FreezableLookup::default().freeze();
+    /// assert_eq!(lookup.lookup("hello"), 0);
+    /// let thawed = lookup.thaw();
+    /// assert_eq!(thawed.lookup("hello"), 1);
+    /// ```
+    pub fn thaw(self) -> Self {
+        use self::FreezableLookupInner::*;
+
+        let inner = match self.0 {
+            Frozen(numberer) => Fresh(numberer.into()),
+            fresh => fresh,
+        };
+
+        FreezableLookup(inner)
+    }
+}
+
+impl<T> Default for FreezableLookup<T>
+where
+    T: Eq + Hash,
+{
+    fn default() -> Self {
+        FreezableLookup(FreezableLookupInner::Fresh(RefCell::new(Numberer::new(1))))
+    }
+}
+
+impl<O> LookupLen for FreezableLookup<O>
+where
+    O: Eq + Hash,
+{
+    fn len(&self) -> usize {
+        use self::FreezableLookupInner::*;
+
+        match &self.0 {
+            Fresh(cell) => cell.borrow().len() + 1,
+            Frozen(numberer) => numberer.len() + 1,
+        }
+    }
+}
+
+impl<B, O> LookupValue<B> for FreezableLookup<O>
+where
+    B: ?Sized + Eq + Hash + ToOwned<Owned = O>,
+    O: Borrow<B> + Clone + Eq + Hash,
+{
+    fn lookup(&self, t: &B) -> usize {
+        use self::FreezableLookupInner::*;
+
+        match &self.0 {
+            Fresh(cell) => cell.borrow_mut().add(t),
+            Frozen(numberer) => numberer.number(t).unwrap_or(0),
```

Comment:
Fixing unknown to `0` means that padding positions need explicit assignment of different values (Tensor initializes with `0` by default).

Reply:
I see this as something that the user of lookups should decide, since there are also uses of lookups where there is no padding, such as classes in non-sequence learning tasks. `0` is just a sentinel for 'the value is not present in the lookup'. If this is confusing, maybe lookup should should have `start_at` as well and this method should return `Option` for values that are absent.

Reply:
I'd prefer the verbose option with returning an Option<usize>, as numberer returns an Option anyway. This way the user gets to call `unwrap_or` and decides which value to assign.